### ISSUE TITLE
Customization of the ordering of extensions

### ISSUE THREAD

**Issue by koenbeuk** [dotnet/efcore#26071]
When configuring a DbContext, it's often useful to have control over the order in which extensions gets registered, e.g.:

```csharp
optionsBuilder.UseA();
optionsBuilder.UseB();
optionsBuilder.UseC();
```

Can EF respect the ordering when registering the extensions? That is, first apply services for A, then B and finally C?

In EF Core 6, extension registration is sorted based on the `extension.GetType().FullName`. Before, the extension registration was technically undefined but practically almost always in the declared order (A, B and finally C).

I'm relying on the ordering of extensions in the [Projectables](https://github.com/koenbeuk/EntityFrameworkCore.Projectables) library as it requires the core services (DbProvider) to have been registered. 

Currently a workaround is to place your extension class in namespace that has a relative sort order that is higher than the database provider's extension.

**Comment by AndriySvyryd** [comment]
We can store the order when an option extension is added, but only use it for service registration, not for the service provider cache key

**Comment by koenbeuk** [comment]
Excellent! does this mean that this will get backported into EF Core 6 upon release? @AndriySvyryd @ajcvickers

**Comment by AndriySvyryd** [comment]
It doesn't need to be backported. It's fixed in 6.0.0 (rtm)
